tools/crm_mon_output.c

```diff
@@ -87,6 +87,92 @@ last_changed_string(const char *last_written, const char *user,
     }
 }
 
+static char *
+op_history_string(xmlNode *xml_op, const char *task, const char *interval_ms_s,
+                  int rc, unsigned int mon_ops) {
+    const char *call = crm_element_value(xml_op, XML_LRM_ATTR_CALLID);
+    char *interval_str = NULL;
+    char *buf = NULL;
+
+    if (interval_ms_s && safe_str_neq(interval_ms_s, "0")) {
+        char *pair = pcmk_format_nvpair("interval", interval_ms_s, "ms");
+        interval_str = crm_strdup_printf(" %s", pair);
+        free(pair);
+    }
+
+    if (is_set(mon_ops, mon_op_print_timing)) {
+        char *last_change_str = NULL;
+        char *last_run_str = NULL;
+        char *exec_str = NULL;
+        char *queue_str = NULL;
+
+        const char *value = NULL;
+
+        time_t epoch = 0;
+
+        if ((crm_element_value_epoch(xml_op, XML_RSC_OP_LAST_CHANGE, &epoch) == pcmk_ok)
+            && (epoch > 0)) {
+            char *time = pcmk_format_named_time(XML_RSC_OP_LAST_CHANGE, epoch);
+            last_change_str = crm_strdup_printf(" %s", time);
+            free(time);
+        }
+
+        if ((crm_element_value_epoch(xml_op, XML_RSC_OP_LAST_RUN, &epoch) == pcmk_ok)
+            && (epoch > 0)) {
+            char *time = pcmk_format_named_time(XML_RSC_OP_LAST_RUN, epoch);
+            last_run_str = crm_strdup_printf(" %s", time);
+            free(time);
+        }
+
+        value = crm_element_value(xml_op, XML_RSC_OP_T_EXEC);
+        if (value) {
+            char *pair = pcmk_format_nvpair(XML_RSC_OP_T_EXEC, value, "ms");
+            exec_str = crm_strdup_printf(" %s", pair);
+            free(pair);
+        }
+
+        value = crm_element_value(xml_op, XML_RSC_OP_T_QUEUE);
+        if (value) {
+            char *pair = pcmk_format_nvpair(XML_RSC_OP_T_QUEUE, value, "ms");
+            queue_str = crm_strdup_printf(" %s", pair);
+            free(pair);
+        }
+
+        buf = crm_strdup_printf("(%s) %s:%s%s%s%s%s rc=%d (%s)", call, task,
+                                interval_str ? interval_str : "",
+                                last_change_str ? last_change_str : "",
+                                last_run_str ? last_run_str : "",
+                                exec_str ? exec_str : "",
+                                queue_str ? queue_str : "",
+                                rc, services_ocf_exitcode_str(rc));
+
+        if (last_change_str) {
+            free(last_change_str);
+        }
+
+        if (last_run_str) {
+            free(last_run_str);
+        }
+
+        if (exec_str) {
+            free(exec_str);
+        }
+
+        if (queue_str) {
+            free(queue_str);
+        }
```

Comment:
Not a big deal, but free() can handle NULL, so checking is optional (it does save the function call)